src/common/browser-adapters/browser-adapter-event-manager.ts

```diff
@@ -51,78 +50,90 @@ export class BrowserAdapterEventManager {
             //wrapping application listener promise responses in a 4-minute promise race
             //prevents the service worker going idle before a response is sent
-            await this.promiseFactory.timeout(result, FOUR_MINUTES);
+            return this.promiseFactory.timeout(result, FOUR_MINUTES);
         } else {
             // it is possible that this is the result of a fire and forget listener
             // wrap in 2-minute timeout to ensure it completes during service worker lifetime
-            return this.timeoutFactory.timeoutType === TimeoutType.Alarm
-                ? this.timeoutFactory.createTimeout(
-                      () => Promise.resolve(result),
-                      TWO_MINUTES,
-                      `wrapper-timeout-${Date.now()}`,
-                  )
-                : this.timeoutFactory.createTimeout(() => Promise.resolve(result), TWO_MINUTES);
+            globalThis.setTimeout(() => {
+                Promise.resolve(result);
+            }, TWO_MINUTES);
+            return Promise.resolve(result);
         }
-        return result;
     }
 
-    public processEvent(eventType: string, eventArgs: any[], isDeferred?: boolean): boolean {
-        if (this.eventsToApplicationListenersMapping[eventType]) {
-            this.eventsToApplicationListenersMapping[eventType].forEach(applicationListener => {
-                this.forwardEventToApplicationListener(applicationListener, eventArgs);
-            });
-            return true;
+    private deferEvent(eventDetails: EventDetails): Promise<any> {
+        const deferredResolution = this.promiseFactory.externalResolutionPromise();
+
+        this.deferredEvents.push({
+            deferredResolution: deferredResolution,
+            ...eventDetails,
+        });
+
+        return deferredResolution.promise;
+    }
+
+    public processEvent(eventType: string, eventArgs: any[]): Promise<any> {
+        const applicationListener = this.eventsToApplicationListenersMapping[eventType];
+        if (applicationListener) {
+            return this.forwardEventToApplicationListener(applicationListener, eventArgs);
         } else {
-            if (!isDeferred) {
-                const eventDetails: EventDetails = {
+            const alreadyDeferred = this.deferredEvents.find(
+                eventDetails =>
+                    eventDetails.eventType === eventType && eventDetails.eventArgs === eventArgs,
+            );
+            if (alreadyDeferred) {
+                return alreadyDeferred.deferredResolution.promise;
+            } else {
+                return this.deferEvent({
                     eventType,
                     eventArgs,
-                };
-                this.deferredEvents.push(eventDetails);
+                });
             }
-            return false;
         }
     }
 
     private unregisterAdapterListener(event: Events.Event<any>, eventType: string) {
         event.removeListener(this.adapterListener(eventType));
```

Comment:
`this.adapterListener` isn't memoized, so I don't think this is going to have the desired effect